dex/runner.go

```diff
@@ -31,26 +59,54 @@ func NewStartStopWaiter(runner Runner) *StartStopWaiter {
 // Stop to signal the Runner to stop, followed by WaitForShutdown to allow
 // shutdown to complete.
 func (ssw *StartStopWaiter) Start(ctx context.Context) {
-	ssw.ctx, ssw.cancel = context.WithCancel(ctx)
+	ssw.init(ctx)
 	ssw.wg.Add(1)
 	go func() {
 		ssw.runner.Run(ssw.ctx)
 		ssw.wg.Done()
 	}()
 }
 
-// Stop signals the Runner to quit. It is not blocking; use WaitForShutdown
-// after Stop to allow the Runner to return.
-func (ssw *StartStopWaiter) Stop() {
-	ssw.cancel()
+// WaitForShutdown blocks until the Runner has returned in response to Stop.
+func (ssw *StartStopWaiter) WaitForShutdown() {
+	ssw.wg.Wait()
 }
 
-// On will be true until the Runner Context is canceled.
-func (ssw *StartStopWaiter) On() bool {
-	return ssw.ctx.Err() == nil
+// Connector is any type that implements the Connect method, which will return
+// a connection error, and a channel that can be used to wait on shutdown after
+// context cancellation.
+type Connector interface {
+	Connect(ctx context.Context) (error, *sync.WaitGroup)
 }
 
-// WaitForShutdown blocks until the Runner has returned in response to Stop.
-func (ssw *StartStopWaiter) WaitForShutdown() {
-	ssw.wg.Wait()
+// ConnectionMaster manages a Connector.
+type ConnectionMaster struct {
+	contextManager
+	wg        *sync.WaitGroup
+	connector Connector
+}
+
+// NewConnectionMaster is the constructor for a new ConnectionMaster.
+func NewConnectionMaster(c Connector) *ConnectionMaster {
+	return &ConnectionMaster{
+		connector: c,
+	}
+}
+
+// Start connects the Connector, and returns any initial connection error. Use
+// Disconnect to shut down the Connector.
+func (c *ConnectionMaster) Connect(ctx context.Context) (err error) {
+	c.init(ctx)
+	c.mtx.Lock()
+	err, c.wg = c.connector.Connect(c.ctx)
+	c.mtx.Unlock()
+	return err
+}
+
+// Disconnect closes the connection and waits for shutdown.
+func (c *ConnectionMaster) Disconnect() {
+	c.Stop()
+	c.mtx.RLock()
+	defer c.mtx.RUnlock()
+	c.wg.Wait()
```

Comment:
This lock seems odd. If you are trying to keep possession of the ConnectionMaster through the duration of this function, I think there is a window between c.Stop() and this RLock() where another thread could obtain the lock.